src/peer_digest.cc

```diff
@@ -349,7 +327,6 @@ peerDigestHandleReply(void *data, StoreIOBuffer receivedData)
         return;
     }
 
-    assert(fetch->pd);
```

Comment:
There also is an useless cbdataReferenceValid(fetch) check below which can be removed, I think.

Reply:
I agree that the related code has several problems, but those problems cannot be solved by just removing that cbdataReferenceValid(fetch) check from the loop condition. Is the corresponding "Some of the calls we make may try to destroy the fetch structure" claim bogus? I am not sure it is bogus because functions called on each loop iterations do call peerDigestFetchedEnough(), which may destroy the fetch structure.

Reply:
No, I meant another one, `if (cbdataReferenceValid(fetch))`, just before the end of the function.

Reply:
I believe my comments/concern apply to both calls: The fetch structure might become invalid/destroyed during a loop iteration, but the code is not careful about dereferencing the fetch pointer.